### ISSUE TITLE
is dnstap position-dependent in Corefile?

### ISSUE THREAD

**Issue by jpmens** [coredns/coredns#1041]
I have the following Corefile:

```
example.aa {
    health :8080
    file zone.db {
        transfer to *
        transfer to 192.168.1.130:5354
    }
}

. {
    dnstap /tmp/dnstap.sock full
    proxy . 192.168.1.10:53
}
```

I notice that all queries are logged to the tap **except** those for zone `example.aa`. 

1. Is the position of the `dnstap` stanza relevant (i.e. "from here forwards") in the Corefile?
2. How can dnstap be configured globally, irrespective of its location in Corefile?

**Comment by johnbelamaric** [comment]
This is expected. By defining the two different server blocks, you have created separate middleware chains for example.aa and .

See https://coredns.io/2017/06/08/how-queries-are-processed-in-coredns/

**Comment by jpmens** [comment]
I am then understanding that there is no possibility to configure, say, `dnstap` globally.

**Comment by jpmens** [comment]
I shouldn't have closed this.

If I configure `dnstap /tmp/dnstap.sock full` in more than one block, `coredns` hangs upon starting.

**Comment by 2buisson** [comment]
I submitted a PR to dnstap so it can listen multiple connections, meanwhile `dnstap tcp://127.0.0.1:6000 full` seems to work in more than one block.

**Comment by jpmens** [comment]
Oh, you're saying coredns was "hanging" because _dnstap_ didn't accept the connection? I didn't think of verifying that. My apologies.

**Comment by 2buisson** [comment]
The hanging is still an issue and in the end I think the issue is on both sides but can be patched in CoreDNS as a start, thanks for reporting. Will close when fixed.

**Comment by jpmens** [comment]
Adding this to prevent closure: has it been fixed or not?

**Comment by miekg** [comment]
With a Corefile of
~~~
example.com. { 
    forward . 8.8.8.8
    log
    dnstap /tmp/dnstap.sock full
}

. { 
    forward . 8.8.8.8
    log
    dnstap /tmp/dnstap.sock full
}
~~~
CoreDNS starts with
~~~
2018/07/21 17:23:31 [ERROR] plugin/dnstap: No connection to dnstap endpoint
2018/07/21 17:23:31 [ERROR] plugin/dnstap: No connection to dnstap endpoint
example.com.:1053
.:1053
2018/07/21 17:23:31 [INFO] CoreDNS-1.2.0
2018/07/21 17:23:31 [INFO] linux/amd64, go1.10.3, 
CoreDNS-1.2.0
linux/amd64, go1.10.3, 
~~~
So this seems to have been fixed.

**Comment by 2buisson** [comment]
I will investigate this in a week or so

**Comment by miekg** [comment]
Thanks @varyoo

**Comment by 2buisson** [comment]
This is working fine, even when dnstap is listening.

A [new test case](https://github.com/dnstap/golang-dnstap/pull/19) in the dnstap tool shows that it can accept multiple connections.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] is dnstap pos..." ]
>Closed #1041.

thanks for the follow up!
